src/go/rdctl/cmd/snapshot.go

```diff
@@ -1,15 +1,142 @@
+			}
+		}
+		return nil
+	},
 }
 
 func init() {
 	rootCmd.AddCommand(snapshotCmd)
 }
+
+func stopBackendCallFuncAndRestartBackend(wrappedFunction cobraFunc) cobraFunc {
+	return func(cmd *cobra.Command, args []string) error {
+		paths, err := paths.GetPaths()
+		if err != nil {
+			return fmt.Errorf("failed to get paths: %w", err)
+		}
+		if err := createSnapshotLock(paths); err != nil {
+			return err
+		}
+		defer removeSnapshotLock(paths)
+
+		connectionInfo, err := config.GetConnectionInfo()
+		if errors.Is(err, os.ErrNotExist) {
+			// If we cannot get connection info from config file (and it
+			// is not specified by the user) then assume main process is
+			// not running.
+			return wrappedFunction(cmd, args)
+		} else if err != nil {
+			return fmt.Errorf("failed to get connection info: %w", err)
+		}
+
+		// Ensure backend is running if the main process is running at all
+		rdClient := client.NewRDClient(connectionInfo)
+		state, err := rdClient.GetBackendState()
+		if errors.Is(err, syscall.ECONNREFUSED) {
+			// If we cannot connect to the server, assume that the main
+			// process is not running.
+			return wrappedFunction(cmd, args)
+		} else if err != nil {
+			return fmt.Errorf("failed to get backend state: %w", err)
+		}
+		if state.VMState != "STARTED" {
+			return errors.New("Rancher Desktop must be fully running or fully shut down to do this action")
+		}
+
+		// Stop and lock the backend
+		desiredState := client.BackendState{
+			VMState: "STOPPED",
+			Locked:  true,
+		}
+		if err := rdClient.PutBackendState(desiredState); err != nil {
+			return fmt.Errorf("failed to stop backend: %w", err)
+		}
+		if err := waitForVMState(rdClient, "STOPPED"); err != nil {
+			return fmt.Errorf("error waiting for backend to stop: %w", err)
+		}
+
+		functionErr := wrappedFunction(cmd, args)
+
+		// Start and unlock the backend
+		desiredState = client.BackendState{
+			VMState: "STARTED",
+			Locked:  false,
+		}
+		startVMErr := rdClient.PutBackendState(desiredState)
+		waitForStartedErr := waitForVMState(rdClient, "STARTED")
+		return errors.Join(functionErr, startVMErr, waitForStartedErr)
+	}
+}
+
+func waitForVMState(rdClient client.RDClient, desiredState string) error {
+	interval := 1 * time.Second
+	numIntervals := 120
+	for i := 0; i < numIntervals; i = i + 1 {
+		state, err := rdClient.GetBackendState()
+		if err != nil {
+			return fmt.Errorf("failed to get backend state: %w", err)
+		}
+		if state.VMState == desiredState {
+			return nil
+		}
+		time.Sleep(interval)
+	}
+	return fmt.Errorf("timed out waiting for backend state %q", desiredState)
+}
+
+func createSnapshotLock(paths paths.Paths) error {
+	if err := os.MkdirAll(paths.AppHome, 0o755); err != nil {
+		return fmt.Errorf("failed to create snapshot lock parent directory: %w", err)
+	}
+	lockPath := filepath.Join(paths.AppHome, "snapshot.lock")
+	file, err := os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL, 0o644)
```

Comment:
Did you mean to open it locked for writing? (That is, make it so other processes cannot re-open the file; think `flock`.)

In general, file locking is not great cross-platform; I believe the best idea would be:
- On *nix: bind to a unix socket (I believe you can't re-bind to an already bound file), or as mentioned use `flock` (though that has fun issues over NFS, not that I think it'll be a problem for us)
- On Windows: Probably a [named mutex](https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-createmutexw).

I'm not sure we need that as a first go, though; if we do want to land that, it might be better as a follow-up PR.

Reply:
The idea here was to do the equivalent of a `touch` but throwing an error if the file already exists. I wanted to use the presence of the file to indicate a locked state... but of course whatever is looking at the lock has to honor it as that. I *think* this scheme will work cross-platform.

I looked at other options than `os.OpenFile` and they all seem to either truncate the file  (`os.Create`, `os.WriteFile`) or to return an error if the file does not already exist (`os.Open`). I may have missed something though.

[The docs for `os.OpenFile` flags](https://pkg.go.dev/os#pkg-constants) indicate what `os.O_EXCL` etc mean.

Reply:
Yeah, I just didn't like that we can easily end up crashing halfway (assume power loss) and have a stale lock. I guess that's why we have a `rdctl unlock`, but that's not great… ah well. Follow-up, if we ever fix it.